etl/steps/data/garden/migration/2024-11-19/migration_distance.py

```diff
@@ -0,0 +1,131 @@
+"""Load a meadow dataset and create a garden dataset."""
+
+import geopandas as gpd
+import pandas as pd
+from owid.catalog import License, Origin
+
+from etl.data_helpers import geo
+from etl.helpers import PathFinder, create_dataset
+
+# Get paths and naming conventions for current step.
+paths = PathFinder(__file__)
+
+# GeoJson file with country boundaries
+URL_TINY_CTY = "https://d2ad6b4ur7yvpq.cloudfront.net/naturalearth-3.3.0/ne_50m_admin_0_tiny_countries.geojson"
+
+NATURAL_EARTH_ORIGIN = Origin(
+    producer="Natural Earth",
+    title="Admin 0 - Countries and Tiny Country Points",
+    url_main="https://www.naturalearthdata.com/",
+    url_download="https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/110m/cultural/ne_110m_admin_0_countries.zip",
+    description="""Natural Earth is a public domain map dataset available at 1:10m, 1:50m, and 1:110 million scales. It features tightly integrated vector and raster data, was built through a collaboration of many volunteers and is supported by NACIS (North American Cartographic Information Society).""",
+    license=License(name="Public Domain", url="https://www.naturalearthdata.com/about/terms-of-use/"),
+    date_accessed="2024-11-19",
+    citation_full="Made with Natural Earth. Free vector and raster map data @ naturalearthdata.com.",
+)
+
+
+def run(dest_dir: str) -> None:
+    #
+    # Load inputs.
+    #
+    # Load meadow dataset.
+    ds_meadow = paths.load_dataset("migrant_stock")
+
+    # Read table from meadow dataset.
+    tb = ds_meadow["migrant_stock_dest_origin"].reset_index()
+
+    # read natural earth data
+    world = gpd.read_file(gpd.datasets.get_path("naturalearth_lowres"))  # type: ignore
+    tiny_countries = gpd.read_file(URL_TINY_CTY)
+
+    # use web mercator projection
+    world = world.to_crs("EPSG:3857")[["name", "geometry"]]
```

Comment:
This is really tricky, and map projections are a minefield, but I think the Web Mercator projection might not be the best for calculating distances as there is quite a lot of distortion in it, especially close to the poles. 

ChatGPT says that using EPSG:4326 would be more appropriate, and then the distances can be calculated using geodesic distances. 

It suggests this instead:

> from shapely.ops import nearest_points
> from geopy.distance import geodesic
> import pandas as pd
> 
> # Use WGS84 (EPSG:4326) for geographic distances
> world = world.to_crs("EPSG:4326")[["name", "geometry"]]
> tiny_countries = tiny_countries.to_crs("EPSG:4326")[["admin", "geometry"]]
> tiny_countries = tiny_countries.rename(columns={"admin": "name"})
> 
> # Add tiny countries to world
> world = pd.concat([world, tiny_countries], ignore_index=True)
> 
> # Harmonize country names
> world = geo.harmonize_countries(
>     df=world,
>     country_col="name",
>     countries_file=paths.country_mapping_path,
> )
> 
> # Calculate geodesic distance matrix using borders
> def calculate_distance_matrix(world):
>     # Create an empty distance matrix
>     distance_matrix = pd.DataFrame(index=world["name"], columns=world["name"])
> 
>     for i, row1 in world.iterrows():
>         for j, row2 in world.iterrows():
>             if i == j:
>                 distance_matrix.iloc[i, j] = 0  # Distance to itself
>             else:
>                 # Get the nearest points between two geometries
>                 point1, point2 = nearest_points(row1.geometry, row2.geometry)
> 
>                 # Calculate geodesic distance between the nearest points
>                 distance_matrix.iloc[i, j] = geodesic(
>                     (point1.y, point1.x), (point2.y, point2.x)
>                 ).kilometers
> 
>     return distance_matrix
> 
> distance_matrix = calculate_distance_matrix(world)

Reply:
I've had a quick go at rerunning it with this and I think the numbers are a bit more in the region expected. In the current code, I don't think it's dealing with the spherical nature of the earth, e.g., the value for Wallis and Futuna is around 37,000km, basically the world's circumference!

Using the code here above it drops down to around 2250km, which makes more sense as it looks like most [emigrants move to nearby New Caledonia](https://admin.owid.io/admin/variables/999825).

Reply:
Here's a decent explanation on the difference between planar and geodesic differences. which I think applies here:
https://pro.arcgis.com/en/pro-app/latest/tool-reference/spatial-analyst/geodesic-versus-planar-distance.htm